### ISSUE TITLE
SSL Connection to MySQL

### ISSUE THREAD

**Issue by sshark** [jasync-sql/jasync-sql#237]
My client can connect to the MySQL server using the non-SSL connection. However, if the client switched to an SSL connection, the client would throw the `BufferNotFullyConsumedException`. I was using the same configuration,

```
[com.github.jasync.sql.db.mysql.MySQLConnection] MySQLConnection.kt:183 - <mysql-connection-2> Transport failure 
com.github.jasync.sql.db.exceptions.BufferNotFullyConsumedException: Buffer was not fully consumed by decoder, 39 bytes to read, decoder is EOFMessageDecoder and message is EOFMessage
```

It seems to be similar to https://github.com/jasync-sql/jasync-sql/issues/134 
Was this fixed? Thanks

**Comment by oshai** [comment]
Thanks for reporting.
According to the issue you linked it might be an authentication issue under the hood.
If that is the case the fix is probably a proper error message instead of that exception.
For that we will need a way to reproduce the issue (for example by adding a test to [MySQLSSLConnectionSpec](https://github.com/jasync-sql/jasync-sql/blob/6de042b540b65ea76dbc9dd96e31657969ec322a/mysql-async/src/test/java/com/github/jasync/sql/db/mysql/MySQLSSLConnectionSpec.kt)). Please let me know if you can do that or provide any other info.

**Comment by sshark** [comment]
My dev environment based on _Spring Boot 2.4.5, spring-data-r2dbc 1.2.8_, and  _jasync-r2dbc-mysql 1.2.2_. My remote MySQL server version is 5.7.32-log (Azure). However, to be sure, I ran against my local MySQL with similar configurations but with a slight difference in the version, 5.7.34. The connections were ok. I have to find out the differences

**Comment by sshark** [comment]
I used [MySQLSSLConnectionSpec](https://github.com/jasync-sql/jasync-sql/blob/6de042b540b65ea76dbc9dd96e31657969ec322a/mysql-async/src/test/java/com/github/jasync/sql/db/mysql/MySQLSSLConnectionSpec.kt) to extract a log. I have replaced the login name with 'x'-es. Thanks

```
[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:56,877][c.g.j.s.d.m.MySQLConnection] <mysql-connection-3> Connected to db.acme.com/10.0.100.1:3306
[TRACE][db-sql-netty-thread-3][2021-06-09 19:13:56,881][c.g.j.s.d.m.c.MySQLFrameDecoder] [connectionId:<mysql-connection-3>] - Reading message type 10 - (count=1,hasDoneHandshake=false,size=76,isInQuery=false,processingColumns=false,processingParams=false,processedColumns=0,processedParams=0)
0: 0a 35 2e 36 2e 34 37 2e     . 5 . 6 . 4 7 . 
1: 30 00 2d fc 00 00 fe fe     0 . - . . . . . 
2: fe fe fe fe fe fe 00 0f     . . . . . . . . 
3: ab 21 02 00 3f 00 15 00     . ! . . ? . . . 
4: 00 00 00 00 00 00 00 00     . . . . . . . . 
5: 00 fe fe fe fe fe fe fe     . . . . . . . . 
6: fe fe fe fe fe 00 6d 79     . . . . . . m y 
7: 73 71 6c 5f 6e 61 74 69     s q l _ n a t i 
8: 76 65 5f 70 61 73 73 77     v e _ p a s s w 
9: 6f 72 64 00                 o r d . 
Total 76 bytes read
}
[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:56,881][c.g.j.s.d.m.d.HandshakeV10Decoder] Auth plugin data length was 21
[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:56,882][c.g.j.s.d.m.d.HandshakeV10Decoder] handshake message was HandshakeMessage(serverVersion=5.6.47.0, connectionId=64557, seed=[-2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2], serverCapabilities=4172559, characterSet=33, statusFlags=2, authenticationMethod=mysql_native_password)
[TRACE][db-sql-netty-thread-3][2021-06-09 19:13:56,883][c.g.j.s.d.m.c.MySQLOneToOneEncoder] Writing message SSLRequestMessage - 
0: 20 00 00 01 08 aa 0a 00     . . . . . . . . 
1: ff ff ff 00 e0 00 00 00     . . . . . . . . 
2: 00 00 00 00 00 00 00 00     . . . . . . . . 
3: 00 00 00 00 00 00 00 00     . . . . . . . . 
4: 00 00 00 00                 . . . . 
Total 36 bytes read

[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:56,927][i.n.h.s.OpenSsl] netty-tcnative not in the classpath; OpenSslEngine will be unavailable.
[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:56,957][i.n.h.s.JdkSslContext] Default protocols (JDK): [TLSv1.2] 
[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:56,957][i.n.h.s.JdkSslContext] Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA]
[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:57,114][i.n.h.s.u.InsecureTrustManagerFactory] Accepting a server certificate: CN=cr6.southeastasia1-a.control.database.windows.net
[TRACE][db-sql-netty-thread-3][2021-06-09 19:13:57,204][c.g.j.s.d.m.c.MySQLOneToOneEncoder] Writing message HandshakeResponseMessage - 
0: 6a 00 00 02 08 aa 0a 00     j . . . . . . . 
1: ff ff ff 00 e0 00 00 00     . . . . . . . . 
2: 00 00 00 00 00 00 00 00     . . . . . . . . 
3: 00 00 00 00 00 00 00 00     . . . . . . . . 
4: 00 00 00 00 67 6b 61 64     . . . . x x x x 
5: 6d 69 6e 40 76 6d 2d 65     x x x x x x x x 
6: xx xx xx xx xx xx xx xx     x x x x x x x x 
7: xx xx xx 00 14 b8 c7 0c     x x x . . . . . 
8: 6a fa 67 71 a0 26 a8 49     j . g q . & . I 
9: 3c 9c 4f c3 65 e8 3a 31     < . O . e . : 1 
10: 20 xx xx xx xx xx xx 00    . x x x x x x . 
11: 6d 79 73 71 6c 5f 6e 61     m y s q l _ n a 
12: 74 69 76 65 5f 70 61 73     t i v e _ p a s 
13: 73 77 6f 72 64 00           s w o r d . 
Total 110 bytes read

[DEBUG][db-sql-netty-thread-3][2021-06-09 19:13:57,217][i.n.h.s.SslHandler] [id: 0x9412e449, L:/192.168.65.93:38122 - R:vm-e2-dev-mysql.mysql.database.azure.com/40.78.233.2:3306] HANDSHAKEN: protocol:TLSv1.2 cipher suite:TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
[TRACE][db-sql-netty-thread-3][2021-06-09 19:13:57,268][c.g.j.s.d.m.c.MySQLFrameDecoder] [connectionId:<mysql-connection-3>] - Reading message type -2 - (count=2,hasDoneHandshake=true,size=44,isInQuery=false,processingColumns=false,processingParams=false,processedColumns=0,processedParams=0)
0: fe 6d 79 73 71 6c 5f 6e     . m y s q l _ n 
1: 61 74 69 76 65 5f 70 61     a t i v e _ p a 
2: 73 73 77 6f 72 64 00 12     s s w o r d . . 
3: 37 29 58 57 32 5e 1f 01     7 ) X W 2 ^ . . 
4: 7c 41 19 64 1d 27 1e 64     | A . d . ' . d 
5: 6b 56 68 00                 k V h . 
Total 44 bytes read
}
[TRACE][db-sql-netty-thread-3][2021-06-09 19:13:57,268][c.g.j.s.d.m.c.MySQLFrameDecoder] got message type -2
[ERROR][db-sql-netty-thread-3][2021-06-09 19:13:57,270][c.g.j.s.d.m.MySQLConnection] <mysql-connection-3> Transport failure com.github.jasync.sql.db.exceptions.BufferNotFullyConsumedException: Buffer was not fully consumed by decoder, 39 bytes to read, decoder is EOFMessageDecoder and message is EOFMessage
	at com.github.jasync.sql.db.mysql.codec.MySQLFrameDecoder.doDecoding(MySQLFrameDecoder.kt:206)
	at com.github.jasync.sql.db.mysql.codec.MySQLFrameDecoder.handleCommonFlow(MySQLFrameDecoder.kt:169)
	at com.github.jasync.sql.db.mysql.codec.MySQLFrameDecoder.decode(MySQLFrameDecoder.kt:97)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
```

**Comment by oshai** [comment]
@sshark - did you manage to find out why it's not reproduced locally?

**Comment by sshark** [comment]
No, I am not familiar with MySQL byte-level communications. I am guessing some mismatching bytes during communication after login. Perhaps, we can take this offline, and please provide some suggestions on how do I go about it. 

FYI, I can log in using similar configurations with the driver `r2dbc-mysql`.

**Comment by oshai** [comment]
From looking at the log you provided again I think the issues is related to the following warning from [mysql docs](https://dev.mysql.com/doc/internals/en/packet-EOF_Packet.html):
```
Caution
The EOF packet may appear in places where a Protocol::LengthEncodedInteger may appear. You must check whether the packet length is less than 9 to make sure that it is a EOF packet.
```
In our case packet length is `44` bytes.

I am trying to figure out exactly what that means, and how to process the message.

**Comment by oshai** [comment]
I finally got to a conclusion that the message is AuthenticationSwitchRequest which was not implemented yet.
This is based on the content of the message:
```
0: fe 6d 79 73 71 6c 5f 6e     . m y s q l _ n 
1: 61 74 69 76 65 5f 70 61     a t i v e _ p a 
2: 73 73 77 6f 72 64 00 12     s s w o r d . . 
3: 37 29 58 57 32 5e 1f 01     7 ) X W 2 ^ . . 
4: 7c 41 19 64 1d 27 1e 64     | A . d . ' . d 
5: 6b 56 68 00                 k V h . 
Total 44 bytes read
```
And also on r2dbc [driver](https://github.com/mirromutth/r2dbc-mysql/blob/bf4874951e440c8c7ae66f52650918993966d2ad/src/main/java/dev/miku/r2dbc/mysql/message/server/ServerMessageDecoder.java#L371)
Created PR #248.

**Comment by oshai** [comment]
Fix is in version 1.2.3.
